### PR TITLE
[REVIEW] Fix rolling test issue with CI

### PR DESCRIPTION
**Pull request by rgsl888prabhu**
Resolving negative window issue, where start_index and end_index were getting negative value which created problem in cuda kernel. As I have not been able to reproduce the issue by myself, I am still not 100% sure this is the fix for CI failure, but it's a genuine issue.

As per the requirement, if the start_index and end_index should be valid with respect to the index provided, which was missing in earlier implementation. Example of having such scenario is pred_window = 3  and fwd_window = -2 in this case start_index = -3 and end_index = -1. Which are improper. In this case, the element should be null.

### CODE REVIEW

File: cpp/tests/rolling/rolling_test.cpp [hunk 1]
```diff
@@ -234,19 +235,21 @@ class RollingTest : public cudf::test::BaseFixture {
...
 
       
       // aggregate
       size_type count = 0;
-      for (size_type j = start_index; j < end_index; j++) {
-        if (!input.nullable() || cudf::bit_is_set(valid_mask, j)) {
-          val = op(static_cast<OutputType>(in_col[j]), val);
-          count++;
-        }
+      if (start_index >= 0 and end_index >= 0) {
```
Comment (harrism) [inline cpp/tests/rolling/rolling_test.cpp]: Unnecessary, can't be negative.
```suggestion
```

File: cpp/src/rolling/rolling.cu [hunk 2]
```diff
@@ -206,8 +206,8 @@ void gpu_rolling(column_device_view input,
     size_type following_window = following_window_begin[i];
 
     // compute bounds
-    size_type start = max(0, i - preceding_window);
-    size_type end = min(input.size(), i + following_window + 1);
+    size_type start = min(num_rows, max(0, i - preceding_window));
```
Comment (harrism) [inline cpp/src/rolling/rolling.cu]: `num_rows` is undefined here, causing CI to fail to compile.

```suggestion
    size_type start = min(input.size(), max(0, i - preceding_window));
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -78,43 +78,43 @@
...
 - PR #4066 Fixed mismatch with dtype enums
-- PR #4078 Fix joins for when column_in_common input parameter is empty
+- PR #4078 Fix joins for when `column_in_common` input parameter is empty
 - PR #4080 Fix multi-index dask test with sort issue
 - PR #4084 Update Java for removal of CATEGORY type
 - PR #4086 ORC reader: fix potentially incorrect timestamp decoding in the last rowgroup
 - PR #4089 Fix dask groupby mutliindex test case issues in join
 - PR #4097 Fix strings concatenate logic with column offsets
 - PR #4076 All null string entries should have null data buffer
 - PR #4145 Support empty index case in DataFrame._from_table
```
Comment (harrism) [inline CHANGELOG.md]: This seems to be the culprit...
```suggestion
- PR #4145 Support empty index case in `DataFrame._from_table`
```

File: cpp/src/rolling/jit/code/kernel.cpp [hunk 4]
```diff
@@ -78,18 +78,15 @@ void gpu_rolling_new(cudf::size_type nrows,
...
-        OutType val = agg_op::template operate<OutType, InType>(in_col, start_index, count);
-
-        // check if we have enough input samples
-        output_is_valid = (count >= min_periods);
-
-        // store the output value, one per thread
-        if (output_is_valid) {
-          out_col[i] = val;
-        }
+    count = min(nrows, max(0, end_index - start_index));
```
Comment (harrism) [inline cpp/src/rolling/jit/code/kernel.cpp]: Please use the same code as the .cu version of the kernel. This is different.

**Comment by harrism** [comment]
@revans2 I don't even know why you were assigned to review. You aren't in rapidsai/cudf-cpp-codeowners are you?

**Comment by rgsl888prabhu** [comment]
> @revans2 I don't even know why you were assigned to review. You aren't in rapidsai/cudf-cpp-codeowners are you?

@harrism I added @revans2 just to validate the requirements of window and nothing else

**Comment by rgsl888prabhu** [comment]
Created a new PR as this is loaded with several commits for testing.

**Pull request closed without merging**